include/mbedtls/check_config.h

```diff
@@ -203,6 +203,19 @@
 #endif
 #endif
 
+#if defined(MBEDTLS_PSA_CRYPTO_C) && \
+    (defined(MBEDTLS_PSA_BUILTIN_ALG_STREAM_CIPHER) || \
+    defined(MBEDTLS_PSA_BUILTIN_ALG_CTR) || \
+    defined(MBEDTLS_PSA_BUILTIN_ALG_CFB) || \
+    defined(MBEDTLS_PSA_BUILTIN_ALG_OFB) || \
+    defined(MBEDTLS_PSA_BUILTIN_ALG_ECB_NO_PADDING) || \
+    defined(MBEDTLS_PSA_BUILTIN_ALG_CBC_NO_PADDING) || \
+    defined(MBEDTLS_PSA_BUILTIN_ALG_CBC_PKCS7) || \
+    defined(MBEDTLS_PSA_BUILTIN_ALG_CCM_STAR_NO_TAG)) && \
+    !defined(MBEDTLS_CIPHER_C)
+#error "Built-in unauthenticated ciphers are enabled in PSA, but not all prerequisites"
```

Comment:
While this does correctly stop the build here in the problematic configuration, it's actually not very helpful - it's not possible to just `scripts/config.py unset MBEDTLS_PSA_BUILTIN_ALG_CTR` and so on, since they are derived variables.

In other words, as a library user who tried to disable `MBEDTLS_CIPHER_C ` but keep `MBEDTLS_CCM_C` enabled, I'm now told "Built-in unauthenticated ciphers are enabled in PSA, but not all prerequisites` but it's not clear what those prerequisites might be.

Normally, I would look at the condition here in `check_config.h` and adjust the config based on the macros being tests - and if it turns out that the things I want are incompatible, that becomes clear. But this really doesn't help me as a library user

Reply:
Ah good point. Error messages from `check_config.h` are usually terse and not helpful by themselves, but that's not a problem most of the time because you can just look above the line that gave the error in order to get the details - but as you point out, that doesn't really work here.

I'm thinking we should expand on the topic in some other place and then give a pointer to that place in the error message. Candidates:
- the documentation of `MBEDTLS_PSA_CRYPTO_C`;
- `docs/driver-only-builds.txt` but that's quite long, so I think pointing to `MBEDTLS_PSA_CRYPTO_C` is easier.

Unfortunately the details are a bit complex, as `BUILTIN_ALG_xxx` is auto-enabled if `ALG_xxx` is enabled and:
- `ALG_xxx` is not accelerated, or
- a key type compatible with `ALG_xxx` is enabled but not accelerated (except if `xxx` is CCM or GCM).

Perhaps we need a layered approach where the error message points to `MBEDTLS_PSA_CRYPTO_C` which gives an overview and then for details points to https://github.com/Mbed-TLS/mbedtls/pull/8641/files#diff-b6ed7355a8cfb932767f2139329ab12adc46d045ea9951d5e322371e05cee73fR293?

Reply:
Hmm, OK, completely different proposal: instead PSA auto-enables `CIPHER_C` when it needs it, so we don't have to put all this complexity in the user's face. We'll still keep the details of when exactly you can disable `CIPHER_C` in `docs/driver-only-builds.md` where it's easier to explain because it's in context - and only in the face of people who are trying to take advantage of drivers to compile out builtins.

Wdyt?

Reply:
My concern with auto-enabling `CIPHER_C` is that potential code-size savings won't be achieved by people who don't realise exactly what they need to disabled when all they really need is CCM or GCM. There are now enough things to disable that "just disable what you don't need" gets to be cumbersome.

Maybe we should document in the comments for `CIPHER_C`, because the current description "Enable the generic cipher layer" is particularly unhelpful these days.

At a high level we could say "enable if you need the mbedtls_cipher_xxx()` APIs", along with "can be disabled if PSA is enabled and all ciphers except CCM and GCM are enabled", or something like that?

So I guess I prefer your first option ("expand on the topic") but suggesting we do that in the documentation for `MBEDTLS_CIPHER_C`.

How does that sound?

Reply:
When A depends on B, we usually mention that in A's documentation rather than B's... That said, I wouldn't object to documenting at B as long as A's documentation has a pointer to it.

> My concern with auto-enabling `CIPHER_C` is that potential code-size savings won't be achieved by people who don't realise exactly what they need to disable when all they really need is CCM or GCM. There are now enough things to disable that "just disable what you don't need" gets to be cumbersome.

Yeah, I was concerned about that too. It's the usual trade-off of "making it harder to get the damn thing to build, but once it builds you have more confidence it does what you mean" - if it were about security or correctness I wouldn't hesitate for a second of course, but about code size?

Also, as a matter of principle, the fact that PSA Crypto depends on `CIPHER_C` for some things is an implementation detail that can and will change in the future, so I'm not super happy about exposing users to it. But OTOH that's how our historic config system works.

> At a high level we could say "enable if you need the mbedtls_cipher_xxx()` APIs", along with "can be disabled if PSA is enabled and all ciphers except CCM and GCM are enabled", or something like that?

This is ambiguous: does "enabled" mean `PSA_WANT_xxx` or `MBEDTLS_yyy_C` and `MBEDTLS_CIPHER_MODE_zzz`? Documenting this in a way that's unambiguous, correct and complete is not easy - that's why we have `docs/driver-only-builds.md`.

Suggestion: "Can be disabled if either PSA crypto is disabled, or PSA is enabled but no unauthenticated cipher is enabled on the PSA side, or PSA is enabled with unauthenticated ciphers but they're all fully provided by a driver (see `docs/driver-only-builds.md` for details about that case)."

But even then, "no unauthenticated cipher is enabled on the PSA side" is not trivial as there are two cases:
- with `MBEDTLS_PSA_CRYPTO_CONFIG` disabled (default) it means no `MBEDTLS_CIPHER_MODE_xxx` is enabled;
- with `MBEDTLS_PSA_CRYPTO_CONFIG` enabled (non-default, but needed for driver-only builds, in particular, enabled by TF-M) it means no `PSA_WANT_ALG_xxx` where xxx represents an unauthenticated cipher.

Also, that's not complete: if you want to disable `CIPHER_C` currently you also need to disable `PKCS5_C`, `PKCS12_C`, `CMAC_C` and `NIST_KW_C`. Which circles back to my point that we usually document "A needs B" at A's, and I don't think switching to the opposite would make things easier.

Reply:
I hate making things that are that hard to document properly. Unfortunately I'm afraid until we've switched to the crypto config being fully based on `PSA_WANT` macros and additive, we're stuck with some complexity. And here, some more related to the fact that we're doing some things only partially due to lack of time.

Reply:
> if you want to disable `CIPHER_C` currently you also need to disable `PKCS5_C`, ...

Those are discoverable from the `check_config.h` errors quite easily. I was hoping this would do the same for the cipher pieces.

> This is ambiguous: does "enabled" mean `PSA_WANT_xxx` or `MBEDTLS_yyy_C` and `MBEDTLS_CIPHER_MODE_zzz`?

Yes, and I don't really know the answer to that :)  It was more a suggestion as to "this is the kind of thing I would like to be told as a library user"

Documenting this solely in `docs/driver-only-builds.md` seems to be wrong since the built-in CCM and GCM can be used in driver-less builds without `CIPHER_C` providing XXX (where XXX is exactly what I'm looking for from this!)

Reply:
> Documenting this solely in `docs/driver-only-builds.md` seems to be wrong since the built-in CCM and GCM can be used in driver-less builds without `CIPHER_C` providing XXX (where XXX is exactly what I'm looking for from this!)

Ah, excellent point. It's all tied up in my mind because its "this quarter's EPIC" but of course from a user's perspective they're different things!

Reply:
> > My concern with auto-enabling CIPHER_C is that potential code-size savings won't be achieved by people who don't realise exactly what they need to disable when all they really need is CCM or GCM. There are now enough things to disable that "just disable what you don't need" gets to be cumbersome.
>
> Yeah, I was concerned about that too. It's the usual trade-off of "making it harder to get the damn thing to build, but once it builds you have more confidence it does what you mean" - if it were about security or correctness I wouldn't hesitate for a second of course, but about code size?

My take on this is that people who care a lot about code size will start from an empty configuration and enumerate their needs, rather than take the default configuration and remove what they don't need. With that approach, the additive method (declare what you need, and its dependencies are automatically enabled) is easier!

The explicit method is only easier in rare circumstances, for example if you have a code size budget and are trying to provide as much functionality as you can within that budget. That's not a common case, so I think it's reasonable that it's harder and will take some fiddling. It's not even really more fiddling than the additive method: the difference is not how hard it is to find a suitable build, but whether you find problems because `check_config.h` complains or because `size` gives a number that's too large.

Reply:
So is that two in favour of additive/auto-enable and one against?

I'll note that even if we go for additive, we'll probably want some documentation in `mbedtls_config.h` - as Tom notes, `driver-only-builds.md` can't be the only place where this is documented, since removing `CIPHER_C` is not just about drivers. Some modules already have an `Auto-enables` line in their documentation, so `MBEDTLS_PSA_CRYPTO_C` could have:
```
Auto-enables: MBEDTLS_CIPHER_C if some unauthenticated is enabled in PSA (unless it's fully accelerated, see docs/driver-only-builds.md about that).
```

@tom-cosgrove-arm Wdyt?

Reply:
I'm okay with auto-enable as long as there's a back-link to it. I'd like a bit more clarity on the message; have discussed off-line with @mpg

Reply:
Suggested resolution based on the discussion with @tom-cosgrove-arm:
* Remove the check in `check_config.h` and instead auto-enable `CIPHER_C` based on PSA needs in `mbedtls/config_adjust_legacy_from_psa.h`.
* Add an `Auto-enables` comment to the documentation of `PSA_CRYPTO_C` with content like "MBEDTLS_CIPHER_C if any unauthenticated (ie, non-AEAD) cipher is enabled in PSA (unless it's fully accelerated, see docs/driver-only-builds.md about that)."
* Add a back-link in the documentation of `CIPHER_C`, like "Auto-enabled by: MBEDTLS_PSA_CRYPTO_C depending on which ciphers are enabled (see the documentation of that option for details)."